src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/CheckstyleJavaParser.g4

```diff
@@ -0,0 +1,858 @@
+
+formalParameter
+    : mods+=variableModifier* type=typeType[true]
+      variableDeclaratorId[$ctx.mods, $ctx.type]
+    ;
+
+lastFormalParameter
+    : mods+=variableModifier* type=typeType[true] annotations[false]
+      ELLIPSIS variableDeclaratorId[$ctx.mods, $ctx.type]
+    ;
+
+qualifiedName
+    : id (DOT id)*
+    ;
+
+literal
+    : integerLiteral
+    | floatLiteral
+    | textBlockLiteral
+    | CHAR_LITERAL
+    | STRING_LITERAL
+    | LITERAL_TRUE
+    | LITERAL_FALSE
+    | LITERAL_NULL
+    ;
+
+integerLiteral
+    : DECIMAL_LITERAL_LONG
+    | DECIMAL_LITERAL
+    | HEX_LITERAL_LONG
+    | HEX_LITERAL
+    | OCT_LITERAL_LONG
+    | OCT_LITERAL
+    | BINARY_LITERAL_LONG
+    | BINARY_LITERAL
+    ;
+
+floatLiteral
+    : DOUBLE_LITERAL
+    | FLOAT_LITERAL
+    | HEX_DOUBLE_LITERAL
+    | HEX_FLOAT_LITERAL
+    ;
+
+textBlockLiteral
+    : TEXT_BLOCK_LITERAL_BEGIN TEXT_BLOCK_CONTENT TEXT_BLOCK_LITERAL_END
+    ;
+
+annotations[boolean createImaginaryNode]
+    : anno+=annotation*
+    ;
+
+annotation
+    : AT qualifiedName
+      (LPAREN ( elementValuePairs | elementValue )? RPAREN)?
+    ;
+
+elementValuePairs
+    : elementValuePair (COMMA elementValuePair)*
+    ;
+
+elementValuePair
+    : id ASSIGN elementValue
+    ;
+
+elementValue
+    : expression
+    | annotation
+    | elementValueArrayInitializer
+    ;
+
+elementValueArrayInitializer
+    : LCURLY (elementValue (COMMA elementValue)*)? (COMMA)? RCURLY
+    ;
+
+annotationTypeDeclaration[List<ModifierContext> mods]
+    : AT LITERAL_INTERFACE id annotationTypeBody
+    ;
+
+annotationTypeBody
+    : LCURLY (annotationTypeElementDeclaration)* RCURLY
+    ;
+
+annotationTypeElementDeclaration
+    : mods+=modifier* annotationTypeElementRest[$ctx.mods]
+    | SEMI // this is not allowed by the grammar, but apparently allowed by the actual compiler
+    ;
+
+annotationTypeElementRest[List<ModifierContext> mods]
+    : type=typeType[true]
+      ( annotationMethodRest[mods, ((AnnotationFieldContext) _localctx).type]
+      | annotationConstantRest[mods, ((AnnotationFieldContext) _localctx).type]
+      )
+      SEMI                                                                 #annotationField
+    | type=typeType[true]  SEMI                                            #annotationType
+    | classDeclaration[mods] SEMI?                                         #annotationType
+    | recordDeclaration[mods] SEMI?                                        #annotationType
+    | interfaceDeclaration[mods] SEMI?                                     #annotationType
+    | enumDeclaration[mods] SEMI?                                          #annotationType
+    | annotationTypeDeclaration[mods] SEMI?                                #annotationType
+    ;
+
+annotationMethodRest[List<ModifierContext> mods, TypeTypeContext type]
+    : id LPAREN RPAREN cStyleArrDec+=arrayDeclarator* defaultValue?
+    ;
+
+annotationConstantRest[List<ModifierContext> mods, TypeTypeContext type]
```

Comment:
We pass the typeType RuleContext to the appropriate rule to add the `TYPE` AST, similar to how we do it in the ANTLR2 grammar.